components/places/src/api/matcher.rs

```diff
@@ -414,6 +444,9 @@ impl<'query> Matcher for OriginOrUrl<'query> {
                 ],
                 SearchResult::from_url_row,
             )?
+            .into_iter()
```

Comment:
This will collect the vec twice, we should just query_and_then_named (or whatever) which will produce an iterator that we can collect a single time.

Goes for the other cases too.

Reply:
@thomcc do you mind helping my poor brain here? :) Note the `.flatten().flatten()` in my most recent commit, and I'm struggling to work out exactly how to abort on Err() rather than ignoring them (TBH, ignoring them probably isn't too bad - but if we did that, we wouldn't need the new Option, so either way, this is wrong!)

Reply:
> TBH, ignoring them probably isn't too bad

Thinking more, that's possibly the cleanest option here - the only other possible errors are from sqlite and/or easily covered by tests. In this context, `Result<Option<T>>` isn't adding value over  `Result<T>` where `Err` is just logged and ignored?

Reply:
This does it. There might be a way to keep the generic Error bound but I'm not sure it was serving much purpose anyway (thats the reason this is so large, the meat of it is `Ok(iter.map(|r| r.transpose()).flatten().collect::<Result<_>>()?)`).

```patch
diff --git a/components/places/src/api/matcher.rs b/components/places/src/api/matcher.rs
index 1dacdfed..ed86298e 100644
--- a/components/places/src/api/matcher.rs
+++ b/components/places/src/api/matcher.rs
@@ -11,22 +11,21 @@ use sql_support::{maybe_log_plan, ConnExt};
 use url::Url;
 
 // A helper to log, cache and execute a query, returning a vector of flattened rows.
-fn query_flat_rows_and_then_named<T, E, F>(
+fn query_flat_rows_and_then_named<T, F>(
     conn: &PlacesDb,
     sql: &str,
     params: &[(&str, &dyn ToSql)],
     mapper: F,
 ) -> Result<Vec<T>>
 where
-    E: From<rusqlite::Error>,
-    F: FnMut(&Row<'_>) -> std::result::Result<Option<T>, E>,
+    F: FnMut(&Row<'_>) -> Result<Option<T>>,
 {
     maybe_log_plan(conn, sql, params);
     let mut stmt = conn.prepare_maybe_cached(sql, true)?;
     let iter = stmt.query_and_then_named(params, mapper)?;
     // XXX - one flatten() is for Result<>, the other for Option<> - markh
     // can't work out how to fail on error!
-    Ok(iter.flatten().flatten().collect())
+    Ok(iter.map(|r| r.transpose()).flatten().collect::<Result<_>>()?)
 }
 
 #[derive(Debug, Clone)]
@@ -204,7 +203,7 @@ pub struct SearchResult {
 impl SearchResult {
     /// Default search behaviors from Desktop: HISTORY, BOOKMARK, OPENPAGE, SEARCHES.
     /// Default match behavior: MATCH_BOUNDARY_ANYWHERE.
-    pub fn from_adaptive_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Option<Self>> {
+    pub fn from_adaptive_row(row: &rusqlite::Row<'_>) -> Result<Option<Self>> {
         let mut reasons = vec![MatchReason::PreviousUse];
 
         let search_string = row.get::<_, String>("searchString")?;
@@ -240,7 +239,7 @@ impl SearchResult {
         })
     }
 
-    pub fn from_suggestion_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Option<Self>> {
+    pub fn from_suggestion_row(row: &rusqlite::Row<'_>) -> Result<Option<Self>> {
         let mut reasons = vec![MatchReason::Bookmark];
 
         let search_string = row.get::<_, String>("searchString")?;
@@ -272,7 +271,7 @@ impl SearchResult {
         })
     }
 
-    pub fn from_origin_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Option<Self>> {
+    pub fn from_origin_row(row: &rusqlite::Row<'_>) -> Result<Option<Self>> {
         let search_string = row.get::<_, String>("searchString")?;
         let url = row.get::<_, String>("url")?;
         let display_url = row.get::<_, String>("displayURL")?;
@@ -294,7 +293,7 @@ impl SearchResult {
         })
     }
 
-    pub fn from_url_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Option<Self>> {
+    pub fn from_url_row(row: &rusqlite::Row<'_>) -> Result<Option<Self>> {
         let search_string = row.get::<_, String>("searchString")?;
         let href = row.get::<_, String>("url")?;
         let stripped_url = row.get::<_, String>("strippedURL")?;
```